heft-plugins/heft-webpack5-plugin/src/Webpack5Plugin.ts

```diff
@@ -0,0 +1,331 @@
+      }
+    }
+    return this._webpackConfiguration;
+  }
+
+  private async _runWebpackAsync(
+    taskSession: IHeftTaskSession,
+    heftConfiguration: HeftConfiguration,
+    webpackConfiguration: IWebpackConfiguration | null,
+    serveMode: boolean,
+    watchMode: boolean
+  ): Promise<void> {
+    if (!webpackConfiguration) {
+      return;
+    }
+
+    const logger: IScopedLogger = taskSession.logger;
+    const webpack: typeof TWebpack = await this._getWebpackAsync();
+    logger.terminal.writeLine(`Using Webpack version ${webpack.version}`);
+
+    let compiler: TWebpack.Compiler | TWebpack.MultiCompiler;
+    if (Array.isArray(webpackConfiguration)) {
+      if (webpackConfiguration.length === 0) {
+        logger.terminal.writeLine('The webpack configuration is an empty array - nothing to do.');
+        return;
+      } else {
+        compiler = webpack.default(webpackConfiguration); /* (webpack.Compilation[]) => MultiCompiler */
+      }
+    } else {
+      compiler = webpack.default(webpackConfiguration); /* (webpack.Compilation) => Compiler */
+    }
+
+    if (serveMode) {
+      const defaultDevServerOptions: TWebpackDevServer.Configuration = {
+        host: 'localhost',
+        devMiddleware: {
+          publicPath: '/',
+          stats: {
+            cached: false,
+            cachedAssets: false,
+            colors: heftConfiguration.terminalProvider.supportsColor
+          }
+        },
+        client: {
+          logging: 'info'
+        },
+        port: 8080
+      };
+
+      let options: TWebpackDevServer.Configuration;
+      if (Array.isArray(webpackConfiguration)) {
+        const devServerOptions: TWebpackDevServer.Configuration[] = webpackConfiguration
+          .map((configuration) => configuration.devServer)
+          .filter((devServer): devServer is TWebpackDevServer.Configuration => !!devServer);
+        if (devServerOptions.length > 1) {
+          logger.emitWarning(
+            new Error(`Detected multiple webpack devServer configurations, using the first one.`)
+          );
+        }
+
+        if (devServerOptions.length > 0) {
+          options = { ...defaultDevServerOptions, ...devServerOptions[0] };
+        } else {
+          options = defaultDevServerOptions;
+        }
+      } else {
+        options = { ...defaultDevServerOptions, ...webpackConfiguration.devServer };
+      }
+
+      // Register a plugin to callback after webpack is done with the first compilation
+      // so we can move on to post-build
+      let firstCompilationDoneCallback: (() => void) | undefined;
+      const originalBeforeCallback: typeof options.setupMiddlewares | undefined = options.setupMiddlewares;
+      options.setupMiddlewares = (middlewares, devServer) => {
+        compiler.hooks.done.tap(PLUGIN_NAME, () => {
+          if (firstCompilationDoneCallback) {
+            firstCompilationDoneCallback();
+            firstCompilationDoneCallback = undefined;
+          }
+        });
```

Comment:
This has no need to be inside of `setupMiddlewares`; none of the callback parameters are used.